CoreServices/WindowServer/libevdev/tools/libevdev-tweak-device.1
.TH LIBEVDEV-TWEAK-DEVICE "1"
.SH NAME
libevdev-tweak-device \- modify an evdev kernel device
.SH SYNOPSIS
.B libevdev-tweak-device
--abs ABS_X [--min a] [--max b] [--res c] [--fuzz d] [--flat e]
/dev/input/eventX
.B libevdev-tweak-device
--resolution res[,yres] /dev/input/eventX
.PP
.B libevdev-tweak-device
--led LED_NUML --on|--off /dev/input/eventX
.SH DESCRIPTION
.PP
The
.I libevdev-tweak-device
tool changes the properties of the evdev kernel device at
.I /dev/input/eventX.
Currently this may be used to force an LED on or off, or to change the
properties of an absolute axis (e.g. its minimum/maximum range or
resolution). Changes are permanent until the device is removed.
.SH OPTIONS
.SS Changing absolute axes
.TP 8
.B --abs axis
Change the given named ABS_ kernel axis, e.g. ABS_X. For a full list, see linux/input.h.
Each of the options
.B min, max, res, fuzz, flat
may be given.
.TP 8
.B --min v
Set the absinfo minimum to the value v
.TP 8
.B --max v
Set the absinfo maximum to the value v
.TP 8
.B --res v
Set the absinfo resolution to the value v
.TP 8
.B --fuzz v
Set the absinfo fuzz to the value v
.TP 8
.B --flat v
Set the absinfo flat to the value v
.PP
.SS Changing the x/y resolution
.TP 8
.B --resolution res[,yres]
Changes the resolution of the ABS_X, ABS_MT_POSITION_X, ABS_Y, and
ABS_MT_POSITION_Y axis to the given resolution. If only one resolution value
is provided, both x and y axis are set to the same resolution, otherwise the
first resolution value is applied to the x axes and the second value to the
y axes.
.SS Toggling LEDs
.TP 8
.B --led led
Change the given LED, e.g. LED_NUML. For a full list, see linux/input.h.
.TP 8
.B --on
Change the LED state to on
.TP 8
.B --off
Change the LED state to off
.SH NOTES
.PP
The kernel does not notify processes about absinfo property changes. Any
process that has previously obtained the absinfo from the device will remain
on the old information. This makes using this tool potentially racy, use
with caution.


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

lib/libevdev/tools/libevdev-tweak-device.c
// SPDX-License-Identifier: MIT
/*
 * Copyright © 2014 Red Hat, Inc.
 */

#include "config.h"

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <libgen.h>
#include <limits.h>
#include <linux/input.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include "libevdev/libevdev.h"

static void
usage(const char *progname)
{
	printf("%s --abs <axis> [--min min] [--max max] [--res res] [--fuzz fuzz] [--flat flat] /dev/input/eventXYZ\n"
	       "\tChange the absinfo struct for the named axis\n"
	       "%s --resolution res[,yres] /dev/input/eventXYZ\n"
	       "\tChange the x/y resolution on the given device\n"
	       "%s --led <led> --on|--off /dev/input/eventXYZ\n"
	       "\tEnable or disable the named LED\n",
	       progname,
	       progname,
	       progname);
}

enum mode {
	MODE_NONE = 0,
	MODE_ABS,
	MODE_LED,
	MODE_RESOLUTION,
	MODE_HELP,
};

enum opts {
	OPT_ABS = 1 << 0,
	OPT_MIN = 1 << 1,
	OPT_MAX = 1 << 2,
	OPT_FUZZ = 1 << 3,
	OPT_FLAT = 1 << 4,
	OPT_RES = 1 << 5,
	OPT_LED = 1 << 6,
	OPT_ON = 1 << 7,
	OPT_OFF = 1 << 8,
	OPT_RESOLUTION = 1 << 9,
	OPT_HELP = 1 << 10,
};

static bool
parse_resolution_argument(const char *arg, int *xres, int *yres)
{
	int matched;

	matched = sscanf(arg, "%d,%d", xres, yres);

	switch(matched) {
		case 2:
			break;
		case 1:
			*yres = *xres;
			break;
		default:
			return false;
	}

	return true;
}

static inline bool
safe_atoi(const char *str, int *val)
{
        char *endptr;
        long v;

        v = strtol(str, &endptr, 10);
        if (str == endptr)
                return false;
        if (*str != '\0' && *endptr != '\0')
                return false;

        if (v > INT_MAX || v < INT_MIN)
                return false;

        *val = v;
        return true;
}

static int
parse_event_code(int type, const char *str)
{
	int code;

	code = libevdev_event_code_from_name(type, str);
	if (code != -1)
		return code;

	if (safe_atoi(str, &code))
		return code;

	return -1;
}

static int
parse_options_abs(int argc, char **argv, unsigned int *changes,
		  int *axis, struct input_absinfo *absinfo)
{
	int rc = 1;
	int c;
	int option_index = 0;
	static struct option opts[] = {
		{ "abs", 1, 0, OPT_ABS },
		{ "min", 1, 0, OPT_MIN },
		{ "max", 1, 0, OPT_MAX },
		{ "fuzz", 1, 0, OPT_FUZZ },
		{ "flat", 1, 0, OPT_FLAT },
		{ "res", 1, 0, OPT_RES },
		{ NULL, 0, 0, 0 },
	};

	if (argc < 2)
		goto error;

	optind = 1;
	while (1) {
		c = getopt_long(argc, argv, "h", opts, &option_index);
		if (c == -1)
			break;

		switch (c) {
			case OPT_ABS:
				*axis = parse_event_code(EV_ABS, optarg);
				if (*axis == -1)
					goto error;
				break;
			case OPT_MIN:
				absinfo->minimum = atoi(optarg);
				break;
			case OPT_MAX:
				absinfo->maximum = atoi(optarg);
				break;
			case OPT_FUZZ:
				absinfo->fuzz = atoi(optarg);
				break;
			case OPT_FLAT:
				absinfo->flat = atoi(optarg);
				break;
			case OPT_RES:
				absinfo->resolution = atoi(optarg);
				break;
			default:
				goto error;
		}
		*changes |= c;
	}
	rc = 0;
error:
	return rc;
}

static int
parse_options_led(int argc, char **argv, int *led, int *led_state)
{
	int rc = 1;
	int c;
	int option_index = 0;
	static struct option opts[] = {
		{ "led", 1, 0, OPT_LED },
		{ "on", 0, 0, OPT_ON },
		{ "off", 0, 0, OPT_OFF },
		{ NULL, 0, 0, 0 },
	};

	if (argc < 2)
		goto error;

	optind = 1;
	while (1) {
		c = getopt_long(argc, argv, "h", opts, &option_index);
		if (c == -1)
			break;

		switch (c) {
			case OPT_LED:
				*led = parse_event_code(EV_LED, optarg);
				if (*led == -1)
					goto error;
				break;
			case OPT_ON:
				if (*led_state != -1)
					goto error;
				*led_state = 1;
				break;
			case OPT_OFF:
				if (*led_state != -1)
					goto error;
				*led_state = 0;
				break;
			default:
				goto error;
		}
	}

	rc = 0;
error:
	return rc;
}

static int
parse_options_resolution(int argc, char **argv, int *xres, int *yres)
{
	int rc = 1;
	int c;
	int option_index = 0;
	static struct option opts[] = {
		{ "resolution", 1, 0, OPT_RESOLUTION },
		{ NULL, 0, 0, 0 },
	};

	if (argc < 2)
		goto error;

	optind = 1;
	while (1) {
		c = getopt_long(argc, argv, "h", opts, &option_index);
		if (c == -1)
			break;

		switch (c) {
			case OPT_RESOLUTION:
				if (!parse_resolution_argument(optarg,
							       xres, yres))
					goto error;
				break;
			default:
				goto error;
		}
	}

	rc = 0;
error:
	return rc;
}

static enum mode
parse_options_mode(int argc, char **argv)
{
	int c;
	int option_index = 0;
	static const struct option opts[] = {
		{ "abs", 1, 0, OPT_ABS },
		{ "led", 1, 0, OPT_LED },
		{ "resolution", 1, 0, OPT_RESOLUTION },
		{ "help", 0, 0, OPT_HELP },
		{ NULL, 0, 0, 0 },
	};
	enum mode mode = MODE_NONE;

	if (argc < 2)
		return mode;

	while (mode == MODE_NONE) {
		c = getopt_long(argc, argv, "h", opts, &option_index);
		if (c == -1)
			break;

		switch (c) {
			case 'h':
			case OPT_HELP:
				mode = MODE_HELP;
				break;
			case OPT_ABS:
				mode = MODE_ABS;
				break;
			case OPT_LED:
				mode = MODE_LED;
				break;
			case OPT_RESOLUTION:
				mode = MODE_RESOLUTION;
				break;
			default:
				break;
		}
	}

	if (optind >= argc && mode != MODE_HELP)
		return MODE_NONE;

	return mode;
}

static void
set_abs(struct libevdev *dev, unsigned int changes,
	unsigned int axis, struct input_absinfo *absinfo)
{
	int rc;
	struct input_absinfo abs;
	const struct input_absinfo *a;

	if ((a = libevdev_get_abs_info(dev, axis)) == NULL) {
		fprintf(stderr,
			"Device '%s' doesn't have axis %s\n",
			libevdev_get_name(dev),
			libevdev_event_code_get_name(EV_ABS, axis));
		return;
	}

	abs = *a;
	if (changes & OPT_MIN)
		abs.minimum = absinfo->minimum;
	if (changes & OPT_MAX)
		abs.maximum = absinfo->maximum;
	if (changes & OPT_FUZZ)
		abs.fuzz = absinfo->fuzz;
	if (changes & OPT_FLAT)
		abs.flat = absinfo->flat;
	if (changes & OPT_RES)
		abs.resolution = absinfo->resolution;

	rc = libevdev_kernel_set_abs_info(dev, axis, &abs);
	if (rc != 0)
		fprintf(stderr,
			"Failed to set absinfo %s: %s",
			libevdev_event_code_get_name(EV_ABS, axis),
			strerror(-rc));
}

static void
set_led(struct libevdev *dev, unsigned int led, int led_state)
{
	int rc;
	enum libevdev_led_value state =
		led_state ? LIBEVDEV_LED_ON : LIBEVDEV_LED_OFF;

	if (!libevdev_has_event_code(dev, EV_LED, led)) {
		fprintf(stderr,
			"Device '%s' doesn't have %s\n",
			libevdev_get_name(dev),
			libevdev_event_code_get_name(EV_LED, led));
		return;
	}

	rc = libevdev_kernel_set_led_value(dev, led, state);
	if (rc != 0)
		fprintf(stderr,
			"Failed to set LED %s: %s",
			libevdev_event_code_get_name(EV_LED, led),
			strerror(-rc));
}

static void
set_resolution(struct libevdev *dev, int xres, int yres)
{
	struct input_absinfo abs;

	abs.resolution = xres;
	if (libevdev_has_event_code(dev, EV_ABS, ABS_X))
		set_abs(dev, OPT_RES, ABS_X, &abs);
	if (libevdev_has_event_code(dev, EV_ABS, ABS_MT_POSITION_X))
		set_abs(dev, OPT_RES, ABS_MT_POSITION_X, &abs);

	abs.resolution = yres;
	if (libevdev_has_event_code(dev, EV_ABS, ABS_Y))
		set_abs(dev, OPT_RES, ABS_Y, &abs);
	if (libevdev_has_event_code(dev, EV_ABS, ABS_MT_POSITION_Y))
		set_abs(dev, OPT_RES, ABS_MT_POSITION_Y, &abs);
}

int
main(int argc, char **argv)
{
	struct libevdev *dev = NULL;
	int fd = -1;
	int rc = EXIT_FAILURE;
	enum mode mode;
	const char *path;
	struct input_absinfo absinfo;
	int axis = -1;
	int led = -1;
	int led_state = -1;
	unsigned int changes = 0; /* bitmask of changes */
	int xres = 0,
	    yres = 0;

	mode = parse_options_mode(argc, argv);
	switch (mode) {
		case MODE_HELP:
			rc = EXIT_SUCCESS;
			/* fallthrough */
		case MODE_NONE:
			usage(basename(argv[0]));
			goto out;
		case MODE_ABS:
			rc = parse_options_abs(argc, argv, &changes, &axis,
					       &absinfo);
			break;
		case MODE_LED:
			rc = parse_options_led(argc, argv, &led, &led_state);
			break;
		case MODE_RESOLUTION:
			rc = parse_options_resolution(argc, argv, &xres,
						      &yres);
			break;
		default:
			fprintf(stderr,
				"++?????++ Out of Cheese Error. Redo From Start.\n");
			goto out;
	}

	if (rc != EXIT_SUCCESS)
		goto out;

	if (optind >= argc) {
		rc = EXIT_FAILURE;
		usage(basename(argv[0]));
		goto out;
	}

	path = argv[optind];

	fd = open(path, O_RDWR);
	if (fd < 0) {
		rc = EXIT_FAILURE;
		perror("Failed to open device");
		goto out;
	}

	rc = libevdev_new_from_fd(fd, &dev);
	if (rc < 0) {
		fprintf(stderr, "Failed to init libevdev (%s)\n", strerror(-rc));
		goto out;
	}

	switch (mode) {
		case MODE_ABS:
			set_abs(dev, changes, axis, &absinfo);
			break;
		case MODE_LED:
			set_led(dev, led, led_state);
			break;
		case MODE_RESOLUTION:
			set_resolution(dev, xres, yres);
			break;
		default:
			break;
	}

out:
	libevdev_free(dev);
	if (fd != -1)
		close(fd);

	return rc;
}
